### PR TITLE
Copter: condition_delay accept absolute times

### PR DESCRIPTION
**Pull request by rmackay9**
Enancement to the Condition-Delay mission command so that it accepts UTC times

### CODE REVIEW

**Comment by magicrub** [comment]
Looking good! Shouldn't this go into Rover and Plane too?

File: libraries/AP_Mission/AP_Mission.cpp [hunk 1]
```diff
@@ -176,6 +176,22 @@ void AP_Mission::update()
         return;
     }
 
+    // check if we have an active do command
+    if (!_flags.do_cmd_loaded) {
```
Comment (magicrub) [inline libraries/AP_Mission/AP_Mission.cpp]: isn't this logic already in here? Was it broken?

**Comment by WickedShell** [comment]
Only briefly reviewed so far, will look more at it in a bit but my initial comments are:

Using the GPS library for the current time is bad as the time only updates when the GPS gets a new timestamp, so a intermittent GPS will cause it to miss the timestamp. Use the HAL instead as this is jumped to match GPS time on the first GPS fix. (This allows works better for *nix based autopilots that are working without GPS but might already have a good time base).

The second is do we want to support true arbitrary dates? I think it could be packed in instead of using hours/minutes etc as independent fields. Just my first thoughts.

File: ArduCopter/commands_logic.cpp [hunk 2]
```diff
@@ -915,3 +922,70 @@ void Copter::do_mount_control(const AP_Mission::Mission_Command& cmd)
     camera_mount.set_angle_targets(cmd.content.mount_control.roll, cmd.content.mount_control.pitch, cmd.content.mount_control.yaw);
 #endif
 }
+
+// get milliseconds from now to a target time of week expressed as day, hour, min, sec, ms
+//    match starts from first non-zero value specified.  I.e. specifying day=0, hour=10 will ignore the day and return time until 10am (utc)
+//    to match with sunday, specify day as 7.  To match with midnight specify hour as 24.
+uint32_t Copter::get_ms_until_time_of_week(uint16_t day, uint16_t hour, uint16_t min, uint16_t sec, uint16_t ms) const
```
Comment (magicrub) [inline ArduCopter/commands_logic.cpp]: why is this in copter? Shouldn't it belong in AP_GPS?

File: libraries/AP_GPS/AP_GPS.cpp [hunk 3]
```diff
@@ -449,6 +449,28 @@ AP_GPS::update(void)
 }
 
 /*
+  GPS time of week in days, hours, minutes, seconds, milliseconds
+*/
+void AP_GPS::time_week_utc(uint8_t instance, uint16_t &day, uint16_t &hour, uint16_t &min, uint16_t &sec, uint16_t &ms) const
```
Comment (magicrub) [inline libraries/AP_GPS/AP_GPS.cpp]: Wouldn't this be better to live in AP_GPS_Backend where other utils already are?

**Comment by magicrub** [comment]
Here's PR https://github.com/diydrones/ardupilot/pull/3711 which brings this feature to Plane and Rover

**Comment by DonLakeFlyer** [comment]
Can the CONDITION_DELAY command changes go into the officual mavlink spec?

File: libraries/AP_GPS/AP_GPS.cpp (hunk 3 above)
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS.cpp]: Please see my above note as to why this should all move out of the GPS library all together. On first fix we move the system time to be based on GPS time, but that time will keep counting forwards even if we lose the GPS for some reason.

**Comment by WickedShell** [comment]
Don: I think they should but I'm wondering if we should adjust the format of the way the delays are expressed.

My proposal is to adjust the delay expression to be two fields rather then the current ones. Particularly to move the first one to be GPS week, the second to be seconds in the GPS week. This would require the GCS to help the user with entering/translating times, but would allow a wider range of absolute times to be expressed, and still allowing the same resolution of local time delays (which could be signified by leaving the GPS week blank). This representation won't lose any precision while being transmitted as a float.

**Comment by rmackay9** [comment]
Thanks very much for the review.
I agree with @WickedShell that we could/should move some of the functions out of the GPS library if we can.  The hal.scheduler might be another option as that's where we put other timing functions.  That might make it easier to move the condition-delay feature completely into the AP_Mission library so that it doesn't need to be implemented for each vehicle separately.

@DonLakeFlyer, yes, very happy to update the meaning of the mavlink fields.  I didn't do that immediately but just because I wasn't sure how many changes would be required before this went in mater.

@WickedShell, in this implementation the user can input just partial times.  So it's possible, with additional do-jump commands, to make the vehicle perform an action each hour for example.  I don't know how useful that is but the feature is here.  So if we change it to a more packed time we might lose that.

If the GCSs were involved, we could probably make it easier to set the UTC time.  For example, i think ardupilot sends it's time to the GCS, displaying this somewhere would be helpful.

Thanks again for reviewing.

**Comment by rmackay9** [comment]
Notes from the dev meeting:
1. drop the day
2. add AP_HAL::Util::get_system_clock()
3. move get_time_utc() to AP_Math or AP_Common
4. replace do-condition-delay with a command called something like NAV_WAIT_UNTIL, NAV_WAIT_UTC, or some other TBD name

**Comment by rmackay9** [comment]
Needs rework so closing.

**Comment by gaucho1978** [comment]
So, finally , will the parameters day,hour,minute be settable to "don't care"?

**Pull request closed without merging**